ENIGMAsystem/SHELL/Universal_System/Extensions/FileDropper/FileDropper.cpp

```diff
@@ -0,0 +1,184 @@
+      if (fname != "") fname += "\n";
+      fname += nameVec[i];
+    }
+
+    DragFinish(hDrop);
+  }
+
+  return rc;
+}
+
+static LRESULT CALLBACK SetHook(int nCode, WPARAM wParam, LPARAM lParam) {
+  if (nCode == HC_ACTION) {
+    CWPSTRUCT *pwp = (CWPSTRUCT *)lParam;
+
+    if (pwp->message == WM_KILLFOCUS) {
+      oldProc = (WNDPROC)SetWindowLongPtrW(enigma::hWnd, GWLP_WNDPROC, (LONG_PTR)HookWndProc);
+      UnInstallHook(hook);
+    }
+  }
+
+  return CallNextHookEx(hook, nCode, wParam, lParam);
+}
+
+static HHOOK InstallHook() {
+  hook = SetWindowsHookExW(WH_CALLWNDPROC, (HOOKPROC)SetHook, NULL, GetWindowThreadProcessId(enigma::hWnd, NULL));
+  return hook;
+}
+
+static void file_dnd_apply_filter(string pattern, bool allowfiles, bool allowdirs, bool allowmulti) {
+  if (pattern == "") { pattern = "."; }
+  pattern = string_replace_all(pattern, " ", "");
+  pattern = string_replace_all(pattern, "*", "");
+  std::vector<string> extVec = split_string(pattern, ';');
+  std::vector<string> nameVec = split_string(fname, '\n');
+  std::vector<string>::size_type sz1 = nameVec.size();
+  std::vector<string>::size_type sz2 = extVec.size();
+  fname = "";
+
+  for (std::vector<string>::size_type i2 = 0; i2 < sz2; i2 += 1) {
+    for (std::vector<string>::size_type i1 = 0; i1 < sz1; i1 += 1) {
+      if (extVec[i2] == "." || extVec[i2] == filename_ext(nameVec[i1])) {
+        if (fname != "") fname += "\n";
+        fname += nameVec[i1];
+      }
+    }
+  }
+
+  nameVec = split_string(fname, '\n');
+  sz1 = nameVec.size();
+  fname = "";
+
+  if (allowmulti) {
+    for (std::vector<string>::size_type i = 0; i < sz1; i += 1) {
+      if (allowfiles && file_exists(nameVec[i])) {
+        if (fname != "") fname += "\n";
+        fname += nameVec[i];
+      } else if (allowdirs && directory_exists(nameVec[i])) {
+        if (fname != "") fname += "\n";
+        fname += nameVec[i];
+      }
+    }
+  } else {
+    if (!nameVec.empty()) {
+      if (allowfiles && file_exists(nameVec[0])) {
+        if (fname != "") fname += "\n";
+        fname += nameVec[0];
+      } else if (allowdirs && directory_exists(nameVec[0])) {
+        if (fname != "") fname += "\n";
+        fname += nameVec[0];
+      }
+    }
+  }
+}
+
+namespace enigma_user {
+
+bool file_dnd_get_enabled() {
+  return file_dnd_enabled;
+}
+
+void file_dnd_set_enabled(bool enable) {
+  file_dnd_enabled = enable;
+  DragAcceptFiles(enigma::hWnd, file_dnd_enabled);
+  if (file_dnd_enabled && hook == NULL)
+  InstallHook(); else fname = "";
+}
+
+void file_dnd_set_files(string pattern, bool allowfiles, bool allowdirs, bool allowmulti) {
+  def_pattern = pattern;
+  def_allowfiles = allowfiles;
+  def_allowdirs = allowdirs;
+  def_allowmulti = allowmulti;
+}
+
+string file_dnd_get_files() {
+  if (fname != "") {
+    file_dnd_apply_filter(def_pattern, def_allowfiles, def_allowdirs, def_allowmulti);
+    while (fname.back() == '\n') fname.pop_back();
```

Comment:
I thought it was not possible to put a new line character in a filename? That's why I'm using it as a delimiter, its the most convenient character and prents to a terminal all pretty, putting each file on a new line. Could you explain the science behind what you are suggesting? if i popback to the point of an empty string, that shouldnt crash either because and empty string is not the same thing as a null pointer.

Reply:
A file name can be anything on Unix/Linux.